File: apps/backend/api/serializers/sync.py
"""Flat, cheap serializers for the mobile-v2 delta-sync feeds (doc 04 §3).

These deliberately avoid DRF ``ModelSerializer`` machinery: every feed is built
from a ``.values()``/annotated queryset and each row is turned into a plain dict
by a module-level function. No per-row DB access, no N+1. The output columns
mirror the client's ``remote_*`` tables (doc 02 §1).
"""


def to_ms(dt):
    """UTC datetime -> integer milliseconds since epoch, or None."""
    if dt is None:
        return None
    return int(dt.timestamp() * 1000)


def dominant_hex(raw):
    """Thumbnail.dominant_color is stored as ``"[r, g, b]"``; emit ``#rrggbb``."""
    if not raw:
        return None
    try:
        inner = raw[1:-1]
        r, g, b = (int(x) for x in inner.split(", "))
        return "#%02x%02x%02x" % (r, g, b)
    except (ValueError, TypeError, IndexError):
        return None


def _video_length_ms(raw):
    if not raw:
        return None
    try:
        return int(float(raw) * 1000)
    except (ValueError, TypeError):
        return None


# --------------------------------------------------------------------------- #
# Photo summary
# --------------------------------------------------------------------------- #
# Field list consumed by the delta photo view's ``.values(*PHOTO_VALUES)`` call.
PHOTO_VALUES = (
    "id",
    "image_hash",
    "owner_id",
    "timestamp",
    "added_on",
    "last_modified",
    "video",
    "video_length",
    "rating",
    "hidden",
    "in_trashcan",
    "removed",
    "public",
    "exif_gps_lat",
    "exif_gps_lon",
    "thumbnail__aspect_ratio",
    "thumbnail__dominant_color",
    "search_instance__search_location",
    "owner__favorite_min_rating",
)


def serialize_photo_row(row):
    """Turn one ``.values()`` dict (annotated with ``has_motion``) into a
    flat wire dict matching the client's ``remote_photo`` columns."""
    min_rating = row.get("owner__favorite_min_rating") or 0
    if row["video"]:
        media_type = "video"
    elif row.get("has_motion"):
        media_type = "motion"
    else:
        media_type = "image"
    return {
        "id": str(row["id"]),
        "image_hash": row["image_hash"],
        "owner_id": row["owner_id"],
        "timestamp": to_ms(row["timestamp"]),
        "added_on": to_ms(row["added_on"]),
        "last_modified": to_ms(row["last_modified"]),
        "type": media_type,
        "video_length_ms": _video_length_ms(row["video_length"]),
        "rating": row["rating"],
        "is_favorite": bool(min_rating and row["rating"] >= min_rating),
        "hidden": bool(row["hidden"]),
        "in_trashcan": bool(row["in_trashcan"]),
        "removed": bool(row["removed"]),
        "is_public": bool(row["public"]),
        "aspect_ratio": row["thumbnail__aspect_ratio"],
        "latitude": row["exif_gps_lat"],
        "longitude": row["exif_gps_lon"],
        "search_location": row["search_instance__search_location"] or "",
        "dominant_color": dominant_hex(row["thumbnail__dominant_color"]),
    }


# --------------------------------------------------------------------------- #
# Person
# --------------------------------------------------------------------------- #
PERSON_VALUES = (
    "id",
    "name",
    "kind",
    "face_count",
    "cover_photo__image_hash",
    "last_modified",
)


def serialize_person_row(row):
    return {
        "id": row["id"],
        "name": row["name"],
        "kind": row["kind"],
        "face_count": row["face_count"],
        "cover_photo_hash": row["cover_photo__image_hash"],
        "last_modified": to_ms(row["last_modified"]),
    }


# --------------------------------------------------------------------------- #
# Albums
# --------------------------------------------------------------------------- #
USER_ALBUM_VALUES = (
    "id",
    "title",
    "owner_id",
    "favorited",
    "cover_photo__image_hash",
    "created_on",
    "last_modified",
)


def serialize_user_album_row(row, photo_ids, shared):
    return {
        "id": row["id"],
        "title": row["title"],
        "owner_id": row["owner_id"],
        "favorited": bool(row["favorited"]),
        "shared": int(shared),
        "cover_hash": row["cover_photo__image_hash"],
        "photo_count": len(photo_ids),
        "created_on": to_ms(row["created_on"]),
        "last_modified": to_ms(row["last_modified"]),
        "photo_ids": photo_ids,
    }


AUTO_ALBUM_VALUES = (
    "id",
    "title",
    "timestamp",
    "favorited",
    "last_modified",
)


def serialize_auto_album_row(row, photo_ids, cover_hash):
    return {
        "id": row["id"],
        "title": row["title"],
        "timestamp": to_ms(row["timestamp"]),
        "favorited": bool(row["favorited"]),
        "photo_count": len(photo_ids),
        "cover_hash": cover_hash,
        "last_modified": to_ms(row["last_modified"]),
        "photo_ids": photo_ids,
    }


# Thing / place / tag: mirror only the list (photo_count + cover hashes), not
# membership (doc 02 §1).
def serialize_named_album_row(row, cover_hashes, extra=None):
    out = {
        "id": row["id"],
        "title": row["title"],
        "photo_count": row["photo_count"],
        "cover_hashes": cover_hashes,
        "last_modified": to_ms(row["last_modified"]),
    }
    if extra:
        out.update(extra)
    return out


# --------------------------------------------------------------------------- #
# Sharing surface (shared_user rows)
# --------------------------------------------------------------------------- #
SHARED_USER_VALUES = (
    "id",
    "username",
    "first_name",
    "last_name",
    "avatar",
    "last_modified",
)


def serialize_shared_user_row(row):
    avatar = row.get("avatar")
    return {
        "id": row["id"],
        "username": row["username"],
        "first_name": row["first_name"],
        "last_name": row["last_name"],
        "avatar_url": (f"/media/{avatar}" if avatar else None),
        "last_modified": to_ms(row["last_modified"]),
    }

File: packages/api-client/src/schemas/sync.ts
/**
 * Zod schemas for the mobile-v2 delta-sync feeds (backend `api/views/sync.py`).
 *
 * Every feed shares one envelope:
 *   { v, items, tombstones, next_cursor, total?, server_time }
 *
 * `v` is validated (a version bump is a breaking wire change the client must
 * notice), but every object uses `.passthrough()` so unknown/extra fields the
 * server may add later are tolerated — forward-compat by construction.
 *
 * Item shapes mirror the flat `serialize_*_row` outputs and the client's
 * `remote_*` tables (doc 02 §1). Timestamps are integer ms epoch (or null);
 * `last_modified` is the keyset-cursor ordering key.
 */
import { z } from "zod";

/** Envelope wrapper: parametrised over the per-entity item schema. */
export function syncEnvelope<T extends z.ZodTypeAny>(item: T) {
  return z
    .object({
      v: z.literal(1),
      items: z.array(item),
      // Backend emits `str(entity_id)`; accept numbers too and coerce to string.
      tombstones: z.array(z.coerce.string()).default([]),
      next_cursor: z.string().nullable(),
      total: z.number().optional(),
      server_time: z.string(),
    })
    .passthrough();
}

/* ---- photo -------------------------------------------------------------- */

export const SyncPhotoItem = z
  .object({
    id: z.string(),
    image_hash: z.string(),
    owner_id: z.number(),
    timestamp: z.number().nullable(),
    added_on: z.number().nullable(),
    last_modified: z.number().nullable(),
    type: z.string(),
    video_length_ms: z.number().nullable(),
    rating: z.number(),
    is_favorite: z.boolean(),
    hidden: z.boolean(),
    in_trashcan: z.boolean(),
    removed: z.boolean(),
    is_public: z.boolean(),
    aspect_ratio: z.number().nullable(),
    latitude: z.number().nullable(),
    longitude: z.number().nullable(),
    search_location: z.string().nullable(),
    dominant_color: z.string().nullable(),
  })
  .passthrough();
export type SyncPhotoItem = z.infer<typeof SyncPhotoItem>;

/* ---- person ------------------------------------------------------------- */

export const SyncPersonItem = z
  .object({
    id: z.number(),
    name: z.string().nullable(),
    kind: z.string().nullable(),
    face_count: z.number().nullable(),
    cover_photo_hash: z.string().nullable(),
    last_modified: z.number().nullable(),
  })
  .passthrough();
export type SyncPersonItem = z.infer<typeof SyncPersonItem>;

/* ---- albums ------------------------------------------------------------- */

export const SyncUserAlbumItem = z
  .object({
    id: z.number(),
    title: z.string(),
    owner_id: z.number().nullable(),
    favorited: z.boolean(),
    shared: z.number(), // 0 | 1
    cover_hash: z.string().nullable(),
    photo_count: z.number(),
    created_on: z.number().nullable(),
    last_modified: z.number().nullable(),
    photo_ids: z.array(z.string()).default([]),
  })
  .passthrough();
export type SyncUserAlbumItem = z.infer<typeof SyncUserAlbumItem>;

export const SyncAutoAlbumItem = z
  .object({
    id: z.number(),
    title: z.string(),
    timestamp: z.number().nullable(),
    favorited: z.boolean(),
    photo_count: z.number(),
    cover_hash: z.string().nullable(),
    last_modified: z.number().nullable(),
    photo_ids: z.array(z.string()).default([]),
  })
  .passthrough();
export type SyncAutoAlbumItem = z.infer<typeof SyncAutoAlbumItem>;

/** Thing / tag list mirror (photo_count + cover hashes, no membership). */
export const SyncNamedAlbumItem = z
  .object({
    id: z.number(),
    title: z.string(),
    photo_count: z.number(),
    cover_hashes: z.array(z.string()).default([]),
    last_modified: z.number().nullable(),
  })
  .passthrough();
export type SyncNamedAlbumItem = z.infer<typeof SyncNamedAlbumItem>;

/** Place list mirror — as named, plus the geolocation zoom level. */
export const SyncPlaceAlbumItem = z
  .object({
    id: z.number(),
    title: z.string(),
    photo_count: z.number(),
    cover_hashes: z.array(z.string()).default([]),
    geolocation_level: z.number().nullable(),
    last_modified: z.number().nullable(),
  })
  .passthrough();
export type SyncPlaceAlbumItem = z.infer<typeof SyncPlaceAlbumItem>;

/* ---- sharing surface ---------------------------------------------------- */

export const SyncSharedUserItem = z
  .object({
    id: z.number(),
    username: z.string().nullable(),
    first_name: z.string().nullable(),
    last_name: z.string().nullable(),
    avatar_url: z.string().nullable(),
    last_modified: z.number().nullable(),
  })
  .passthrough();
export type SyncSharedUserItem = z.infer<typeof SyncSharedUserItem>;

/* ---- per-entity envelopes ---------------------------------------------- */

export const SyncPhotosResponse = syncEnvelope(SyncPhotoItem);
export const SyncPersonsResponse = syncEnvelope(SyncPersonItem);
export const SyncUserAlbumsResponse = syncEnvelope(SyncUserAlbumItem);
export const SyncAutoAlbumsResponse = syncEnvelope(SyncAutoAlbumItem);
export const SyncThingAlbumsResponse = syncEnvelope(SyncNamedAlbumItem);
export const SyncPlaceAlbumsResponse = syncEnvelope(SyncPlaceAlbumItem);
export const SyncTagAlbumsResponse = syncEnvelope(SyncNamedAlbumItem);
export const SyncSharingResponse = syncEnvelope(SyncSharedUserItem);

export type SyncPhotosResponse = z.infer<typeof SyncPhotosResponse>;
export type SyncPersonsResponse = z.infer<typeof SyncPersonsResponse>;
export type SyncUserAlbumsResponse = z.infer<typeof SyncUserAlbumsResponse>;
export type SyncAutoAlbumsResponse = z.infer<typeof SyncAutoAlbumsResponse>;
export type SyncThingAlbumsResponse = z.infer<typeof SyncThingAlbumsResponse>;
export type SyncPlaceAlbumsResponse = z.infer<typeof SyncPlaceAlbumsResponse>;
export type SyncTagAlbumsResponse = z.infer<typeof SyncTagAlbumsResponse>;
export type SyncSharingResponse = z.infer<typeof SyncSharingResponse>;

/** Generic envelope shape (for code that is polymorphic over entity). */
export type SyncEnvelope<T> = {
  v: 1;
  items: T[];
  tombstones: string[];
  next_cursor: string | null;
  total?: number;
  server_time: string;
};

/* ---- counts (integrity check, doc 03 §8) ------------------------------- */

export const SyncCounts = z
  .object({
    photos: z.number(),
    persons: z.number(),
    user_albums: z.number(),
    auto_albums: z.number(),
    thing_albums: z.number(),
    place_albums: z.number(),
    tags: z.number(),
    server_time: z.string(),
  })
  .passthrough();
export type SyncCounts = z.infer<typeof SyncCounts>;
